### DISCUSSION TITLE
Proposal: Add `redirect_slashes_internally` flag to Router

### DISCUSSION THREAD

**Discussion by Synrom** [encode/starlette#2769] [Ideas]
## Motivation:
Many users, particularly those using FastAPI (which is built on Starlette), may have encountered a common issue: their application works flawlessly in a local development environment but fails when deployed behind a proxy setup. The reason is temporary redirects, which are triggered by default when a registered route differs from a request only in a trailing slash.

## Proposal:
I propose adding a new flag to the Router class: `redirect_slashes_internally`.

When this flag is enabled, the Router resolves requests with mismatched trailing slashes internally, without issuing a temporary redirect. This change would only require a minor update to the Router.app method, as shown below:
```diff
diff --git a/starlette/routing.py b/starlette/routing.py
index 3b3c529..188c87c 100644
--- a/starlette/routing.py
+++ b/starlette/routing.py
@@ -580,6 +580,7 @@ class Router:
         self,
         routes: typing.Sequence[BaseRoute] | None = None,
         redirect_slashes: bool = True,
+        redirect_slashes_internally: bool = False,
         default: ASGIApp | None = None,
         on_startup: typing.Sequence[typing.Callable[[], typing.Any]] | None = None,
         on_shutdown: typing.Sequence[typing.Callable[[], typing.Any]] | None = None,
@@ -591,6 +592,7 @@ class Router:
     ) -> None:
         self.routes = [] if routes is None else list(routes)
         self.redirect_slashes = redirect_slashes
+        self.redirect_slashes_internally = redirect_slashes_internally
         self.default = self.not_found if default is None else default
         self.on_startup = [] if on_startup is None else list(on_startup)
         self.on_shutdown = [] if on_shutdown is None else list(on_shutdown)
@@ -757,6 +759,10 @@ class Router:
             for route in self.routes:
                 match, child_scope = route.matches(redirect_scope)
                 if match != Match.NONE:
+                    if self.redirect_slashes_internally:
+                        scope.update(child_scope)
+                        await route.handle(scope, receive, send)
+                        return
                     redirect_url = URL(scope=redirect_scope)
                     response = RedirectResponse(url=str(redirect_url))
                     await response(scope, receive, send)
```

## Example:
In the end, you could create an router as follows:
```python
app_resolve_internally = Router(
    routes=[
        Route("/withoutslash", homepage),
        Route("/withslash/", homepage),
    ],
    redirect_slashes_internally=True,
)
```
In this setup, requests to both "/withslash" and "/withoutslash/" are resolved internally

I’d love to hear your thoughts on this proposal :)

**Comment by 7sStudio** [comment]
I can't wait for this issue to be fixed ! It's one of those things that really bothers me with FastAPI

**Reply by Kludex** [reply]
It's not a bug, nothing to be fixed.

**Reply by 7sStudio** [reply]
https://github.com/fastapi/fastapi/pull/12145

**Reply by 7sStudio** [reply]
The lack of a configuration flag is an issue to be fixed in my opinion, not the current behavior.

**Reply by woodgoblin** [reply]
It really messes up contract testing and pretty much all the use cases when you aren't fanatically following the trailing slash rule. OpenAPI is a problem as well. Everyone is working it around.
I can't see why you think it's not a bug.
